File: python/polyomino/polyomino.py
from point import Point

class Polyomino(object):

    def __init__(self, *args):
        if(len(args)) > 1:
            self.points = frozenset(Point(x, y) for (x, y) in args)
        else:
            self.points = frozenset(args[0])

    def add(self, point):
        self.points |= {point}

    def __contains__(self, point):
        return point in self.points

    def upper_left_corner(self):
        x = min(p.x for p in self.points)
        y = min(p.y for p in self.points)
        return Point(x, y)

    def lower_right_corner(self):
        x = max(p.x for p in self.points)
        y = max(p.y for p in self.points)
        return Point(x, y)

    def width(self):
        return self.lower_right_corner().x - self.upper_left_corner().x + 1

    def height(self):
        return self.lower_right_corner().y - self.upper_left_corner().y + 1

    def move(self, dx, dy):
        return Polyomino(p.move(dx, dy) for p in self.points)

    def rotate_right(self, point=Point.origin):
        return Polyomino(p.rotate_right(point) for p in self.points) 

    def rotate_left(self, point=Point.origin):
        return Polyomino(p.rotate_left(point) for p in self.points) 
    
    def reflect_vertically(self, x=0):
        return Polyomino(p.reflect_vertically(x) for p in self.points)

    def reflect_horizontally(self, y=0):
        return Polyomino(p.reflect_horizontally(y) for p in self.points)

    def move_to_origin(self):
        ulc = self.upper_left_corner()
        return self.move(-ulc.x, -ulc.y) 

    def render(self):
        polyomino = self.move_to_origin()

        def render_cell(x, y): 
            return "[]" if Point(x, y) in polyomino.points else "  "

        s = "\n".join(
            "".join(render_cell(x, y) for x in range(self.width()))
            for y in range(self.height())
        )

        return "\n" + s + "\n" 

    def __hash__(self):
        return hash(self.points)

    def __eq__(self, other):
        return self.points == other.points 

    def __cmp__(self, other):
        return cmp(sorted(self.points), sorted(other.points))

    def __repr__(self):
        return "Polyomino(" + str(list(self.points)) + ")" 

    def __str__(self):
        return self.render()

File: rust/polyomino/src/polyomino.rs
#![allow(dead_code)]

use im::OrdSet;
use std::fmt;
use std::iter::successors;

use crate::point::*;

#[derive(Debug, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct Polyomino(OrdSet<Point>);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Dimensions {
    pub width: Coordinate,
    pub height: Coordinate,
}

impl Polyomino {
    pub fn from<const N: usize>(arr: [(Coordinate, Coordinate); N]) -> Self {
        let points = arr.map(|(x, y)| Point::new(x, y));
        Polyomino(OrdSet::from(&points[..]))
    }

    pub fn add_point(&self, p: Point) -> Self {
        Self(self.0.update(p))
    }

    pub fn contains(&self, p: Point) -> bool {
        self.0.contains(&p)
    }

    pub fn shift(&self, dx: Coordinate, dy: Coordinate) -> Self {
        self.map(|p| p.shift(dx, dy))
    }

    pub fn rotate_right_around(&self, rotation_point: Point) -> Self {
        self.map(|p| p.rotate_right_around(rotation_point))
    }

    pub fn rotate_right(&self) -> Self {
        self.rotate_right_around(Point::ORIGIN)
    }

    pub fn rotate_left_around(&self, rotation_point: Point) -> Self {
        self.map(|p| p.rotate_left_around(rotation_point))
    }

    pub fn rotate_left(&self) -> Self {
        self.rotate_left_around(Point::ORIGIN)
    }

    pub fn reflect_vertically(&self, x: Coordinate) -> Self {
        self.map(|p| p.reflect_vertically(x))
    }

    pub fn reflect_over_the_y_axis(&self) -> Self {
        self.reflect_vertically(0)
    }

    pub fn reflect_horizontally(&self, x: Coordinate) -> Self {
        self.map(|p| p.reflect_horizontally(x))
    }

    pub fn reflect_over_the_x_axis(&self) -> Self {
        self.reflect_horizontally(0)
    }

    pub fn upper_left_corner(&self) -> Point {
        let min_x = self.0.iter().map(|p| p.x).min().unwrap();
        let min_y = self.0.iter().map(|p| p.y).min().unwrap();
        Point::new(min_x, min_y)
    }

    pub fn lower_right_corner(&self) -> Point {
        let max_x = self.0.iter().map(|p| p.x).max().unwrap();
        let max_y = self.0.iter().map(|p| p.y).max().unwrap();
        Point::new(max_x, max_y)
    }

    pub fn dimensions(&self) -> Dimensions {
        let ulc = self.upper_left_corner();
        let lrc = self.lower_right_corner();

        Dimensions {
            width: lrc.x - ulc.x + 1,
            height: lrc.y - ulc.y + 1,
        }
    }

    pub fn shift_to_origin(&self) -> Self {
        let point = self.upper_left_corner();
        self.shift(-point.x, -point.y)
    }

    pub fn all_rotations(&self) -> impl Iterator<Item = Polyomino> {
        successors(Some(self.shift_to_origin()), |p| {
            Some(p.rotate_right().shift_to_origin())
        })
        .take(4)
    }

    pub fn all_congruents(&self) -> OrdSet<Polyomino> {
        self.all_rotations()
            .chain(self.reflect_over_the_x_axis().all_rotations())
            .collect()
    }

    pub fn to_canonical_form(&self) -> Polyomino {
        self.all_congruents()
            .into_iter()
            .filter(|p| {
                let d = p.dimensions();
                d.width >= d.height
            })
            .max()
            .unwrap()
    }

    fn map<F: Fn(&Point) -> Point>(&self, f: F) -> Self {
        Polyomino(self.0.iter().map(f).collect())
    }
}

impl fmt::Display for Polyomino {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let polyomino = self.shift_to_origin();
        let d = polyomino.dimensions();
        f.write_str("\n")?;

        for y in 0..d.height {
            for x in 0..d.width {
                if polyomino.contains(Point::new(x, y)) {
                    f.write_str("[]")?
                } else {
                    f.write_str("  ")?
                }
            }

            f.write_str("\n")?
        }

        fmt::Result::Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const POINT_OF_ROTATION: Point = Point::new(3, 7);

    pub fn make_polyomino() -> Polyomino {
        Polyomino::from([(0, 0), (1, 0), (2, 0), (2, 1)])
    }

    #[test]
    pub fn test_rotate_180_oposite_directions() {
        assert_eq!(
            make_polyomino()
                .rotate_right_around(POINT_OF_ROTATION)
                .rotate_right_around(POINT_OF_ROTATION),
            make_polyomino()
                .rotate_left_around(POINT_OF_ROTATION)
                .rotate_left_around(POINT_OF_ROTATION)
        );
    }

    #[test]
    pub fn test_rotate_360() {
        assert_eq!(
            make_polyomino()
                .rotate_right_around(POINT_OF_ROTATION)
                .rotate_right_around(POINT_OF_ROTATION)
                .rotate_right_around(POINT_OF_ROTATION)
                .rotate_right_around(POINT_OF_ROTATION),
            make_polyomino()
        );

        assert_eq!(
            make_polyomino()
                .rotate_left_around(POINT_OF_ROTATION)
                .rotate_left_around(POINT_OF_ROTATION)
                .rotate_left_around(POINT_OF_ROTATION)
                .rotate_left_around(POINT_OF_ROTATION),
            make_polyomino()
        )
    }

    #[test]
    pub fn test_dimensions() {
        assert_eq!(
            make_polyomino().dimensions(),
            Dimensions {
                width: 3,
                height: 2
            }
        );

        assert_eq!(
            make_polyomino().dimensions(),
            make_polyomino()
                .rotate_right_around(POINT_OF_ROTATION)
                .rotate_right_around(POINT_OF_ROTATION)
                .dimensions()
        )
    }

    #[test]
    pub fn test_shift() {
        let p = make_polyomino();

        assert_eq!(p.dimensions(), p.shift(3, 7).dimensions());

        assert_eq!(
            p.shift(33, 37).shift_to_origin().upper_left_corner(),
            Point::ORIGIN
        )
    }

    #[test]
    pub fn test_display() {
        assert_eq!(
            format!("{}", make_polyomino().rotate_right()),
            "
  []
  []
[][]
"
        )
    }

    #[test]
    pub fn test_add_point() {
        assert_eq!(
            make_polyomino().add_point(Point::new(2, 2)),
            Polyomino::from([(0, 0), (1, 0), (2, 0), (2, 1), (2, 2)])
        );
    }

    #[test]
    pub fn test_canonical_form() {
        assert_eq!(
            make_polyomino().rotate_left().shift(12, 17).to_canonical_form(),
            make_polyomino().rotate_right().shift(1, 2).to_canonical_form()
        )
    }
}
